packages/fuels/src/cli/commands/deploy/deployContract.ts

```diff
@@ -25,6 +25,10 @@ export async function deployContract(
   const abi = JSON.parse(readFileSync(abiPath, 'utf-8'));
   const contractFactory = new ContractFactory(bytecode, abi, wallet);
 
-  const contract = await contractFactory.deployContract(deployConfig);
+  const { contract } = await contractFactory.deployContract({
+    awaitExecution: true,
+    ...deployConfig,
+  });
+
```

Comment:
Have you considered splitting the functionality?

I think it could bring some more cohesion and make it easier to follow. 

```ts
// breaking change
const { txResponse, contractId } = contractFactory.deployContract(config);

// new method
const { contract } = await contractFactory.deployAndAwaitForContract(config);
```

To avoid breaking changes, the opposite could be done:

```ts
// same as is
const { contract } = contractFactory.deployContract(config);

// new method
const { txResponse, contractId } = contractFactory.deployContractForLater(config);
```

Reply:
In the current approach, one must wait on a promise that doesn't contain the value they're waiting for.

The value (`Contract` instance) is outside the promise.

Reply:
For combining the Promise and the `Contract` instance, another option could be:

```ts
// breaking change
const { contractId, waitForDeployment } = contractFactory.deployContract(config);

// new method
const { contract } = await waitForDeployment();
```

Reply:
@arboleya Using a secondary method was my initial approach: https://github.com/FuelLabs/fuels-ts/pull/2597/commits/5f6ffdd15a0b9689eed09f6821a1e55934ee007b

@FuelLabs/sdk-ts How do you feel about this proposal?

```ts
// same as is
const { contract } = contractFactory.deployContract(config);

// new method
const { txResponse, contractId } = await contractFactory.deployContractForLater(config);
```

Reply:
I didn't have the context of your first approach, sorry. I like the idea of a second method!

Even on the non-blocking route, we can offer a `contract` instance at the end.

I just want the instance to come from the Promise I'm waiting for:

```ts
const { awaitForDeployment } = await contractFactory.deployForLater(config);
const { contract } = await awaitForDeployment();
```

We could also return the other properties, like `txReponse`, `contractId` etc.:

```ts
const {
  txReponse,
  contractId,
  awaitForDeployment,
} = await contractFactory.deployForLater(config);
```

And it would still be non-breaking.

Reply:
> I just want the instance to come from the Promise I'm waiting for:
```ts
const { awaitForDeployment } = await contractFactory.deployForLater(config);
const { contract } = await awaitForDeployment();
```

I think there is credence to @arboleya suggestion because we would still be enabling non-blocking call, and it is also poka-yoked to prevent users from calling a contract instance that is not yet deployed. 

In the [current state](https://github.com/FuelLabs/fuels-ts/pull/2597/files#diff-a8671e1aac0508f14642db9f3313cc1297d41afed42f68e12448c8973de3989fR32-R36) of the PR, if I call  `factory.deployContract` with `awaitExecution: false`, the `contract` instance returned _may_ throw an error that I will have to handle in my code. I think we should steer consumers away from that code path.

Reply:
I agree with @maschad sentiment. However given the current design, I can't see a way of doing this without making _many_ changes as we can't do this at just the `Contract` level, this would need to be drilled to the invocation scopes. I think the value for users is worth the change. However can we better defend against the breaking change by making default behaviour `awaitExecution: true`?

Reply:
https://github.com/FuelLabs/fuels-ts/pull/2597/commits/c0b06bbc64a49daaae493b3a83cdfc78b56de7df